### PR TITLE
MarkerListActivity: Switch to RecyclerView

### PR DESCRIPTION
**Pull request by lmarz**
Part of: #1613

I've refactored quite a bit. Could you test these changes? I might have missed to reimplement some feature.

### CODE REVIEW

File: src/main/java/de/dennisguse/opentracks/data/ContentProviderUtils.java [hunk 1]
```diff
@@ -548,6 +548,36 @@ private Cursor getMarkerCursor(String[] projection, String selection, String[] s
         return contentResolver.query(MarkerColumns.CONTENT_URI, projection, selection, selectionArgs, sortOrder);
     }
 
+    public List<Marker> searchMarkers(Track.Id trackId, String query) {
+        ArrayList<Marker> markers = new ArrayList<>();
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/data/ContentProviderUtils.java:552]: Create in line 570 (aka short before it is going to be used.

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListActivity.java [hunk 2]
```diff
@@ -157,10 +136,6 @@ protected void onStop() {
     @Override
     protected void onDestroy() {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListActivity.java:137]: onDestroy() is only calling super.
Why?

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListActivity.java [hunk 3]
```diff
@@ -175,23 +150,24 @@ public boolean onCreateOptionsMenu(Menu menu) {
 
         insertMarkerMenuItem = menu.findItem(R.id.marker_list_insert_marker);
 
-        searchMenuItem = menu.findItem(R.id.marker_list_search);
-        ActivityUtils.configureSearchWidget(this, searchMenuItem);
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListActivity.java:179]: Why this change?

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        private final TextView timeDistance;
+        private final TextView time;
+        private View view;
+
+        private Marker marker;
+
+        public ViewHolder(@NonNull View view) {
+            super(view);
+
+            photo = view.findViewById(R.id.marker_list_item_photo);
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java:189]: Can ViewBinding be used?

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                Log.e(TAG, "Failed to image " + e);
+            }
+        });
+    }
+
+    /**
+     * Gets the photo height.
+     */
+    private int getPhotoHeight() {
+        int[] attrs = new int[]{android.R.attr.listPreferredItemHeight};
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java:300]: Sounds like a util function.
ThemeUtils?

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+    /**
+     * It loads the photoUrl in the imageView from view.
+     * It takes the photo from cache or from storage if isn't in the cache.
+     *
+     * @param imageView view object where photo will be loaded.
+     * @param photoUri  photo's url.
+     * @param id        marker's id where photo belong.
+     */
+    private void asyncLoadPhoto(ImageView imageView, Uri photoUri, long id) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java:274]: `long id` can be `Marker.Id id`

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+        Bitmap bm = memoryCache.get(id);
+        imageView.setImageBitmap(bm);
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java:277]: Set only if not null.

**Review by dennisguse — CHANGES_REQUESTED** [review]
Opening marker list per track works.

Crashes for me with a NPE due to Track.Id is null (aka show all markers for all tracks).
I tried to open via Marker icon in TrackListActivity.

PS/ some minor changes

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java (hunk 7 above)
Comment (lmarz) [reply src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java:275]: I wouldn't do that. The View Holder could still be containing an image from a previous list item, that is not visible anymore. When this new Marker doesn't have an image, the old one would still remain

**Review by dennisguse — APPROVED** [review]
Almost there; functionality works :D

Just rebase to main (there is a tiny conflict) and some cleanups and this is good to go!

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+                asyncLoadPhoto(photo, marker.getPhotoURI(), marker.getId());
+            }
+
+            name.setText(marker.getName());
+
+            {
+                ZoneOffset timeZone = new ContentProviderUtils(context).getTrack(marker.getTrackId()).getZoneOffset();
+                OffsetDateTime offsetDateTime = OffsetDateTime.ofInstant(marker.getTime(), timeZone);
+                String dateContent = StringUtils.formatDateTodayRelative(context, offsetDateTime);
```
[the remark this answers is not in the document]
Comment (lmarz) [reply src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java]: I moved it into ListItemUtils. We will also need it for TrackListActivity

File: src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+            if (selection.valueAt(i)) {
+                arrayList.add(selection.keyAt(i));
+            }
+        }
+
+        return arrayList.stream().mapToInt(i -> i).toArray();
+    }
+
+    private long[] getCheckedIds() {
+        List<Long> arrayList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (lmarz) [reply src/main/java/de/dennisguse/opentracks/ui/markers/MarkerListAdapter.java]: It would like to keep it a long, since it's the id of the ViewHolder, which can only be a long

File: src/main/java/de/dennisguse/opentracks/data/ContentProviderUtils.java [hunk 10]
```diff
@@ -548,6 +548,35 @@ private Cursor getMarkerCursor(String[] projection, String selection, String[] s
         return contentResolver.query(MarkerColumns.CONTENT_URI, projection, selection, selectionArgs, sortOrder);
     }
 
+    public List<Marker> searchMarkers(Track.Id trackId, String query) {
+        String selection = null;
+        String[] selectionArgs = null;
+        String sortOrder = null;
+
+        if(query == null) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/data/ContentProviderUtils.java:556]: Just one more thing: formatting. :)

`if(` -> `if (` 
At leasts that what my AndroidStudio is doing on reformat.
All else looks okay.

Comment (lmarz) [reply src/main/java/de/dennisguse/opentracks/data/ContentProviderUtils.java]: Sorry, I always forget to run this command 😅
